feed/script.js

```diff
@@ -0,0 +1,205 @@
+const activityList = document.querySelector('.activity-list');
+const tabsList = document.querySelector('.tabs');
+const lastElementContainer = document.querySelector(LAST_ELEMENT_CONTAINER);
+
+let query = {};
+let newLink = '';
+let activityFeedPage = 0;
+let nextLink = '';
+let isDataLoading = false;
+let category = CATEGORY.ALL;
+
+const tabsData = [
+  { name: 'All', 'data-type': CATEGORY.ALL, class: 'active' },
+  { name: 'Task Requests', 'data-type': CATEGORY.TASK_REQUESTS },
+  { name: 'Extension Requests', 'data-type': CATEGORY.EXTENSION_REQUESTS },
+  { name: 'Task', 'data-type': CATEGORY.TASK },
+  { name: 'OOO Requests', 'data-type': CATEGORY.OOO },
+];
+
+tabsData.forEach((tab) => {
+  const listItem = document.createElement('li');
+  listItem.textContent = tab.name;
+  listItem.dataset.type = tab['data-type'];
+  if (tab.class) {
+    listItem.classList.add(tab.class);
+  }
+  tabsList.appendChild(listItem);
+});
+
+const tabs = document.querySelectorAll('.tabs li');
+
+tabs.forEach((tab) => {
+  tab.addEventListener('click', async function () {
+    tabs.forEach((t) => t.classList.remove('active'));
+    this.classList.add('active');
+    category = this.dataset.type;
+    changeFilter();
+    populateActivityFeed({ category });
+  });
+});
+
+const changeFilter = () => {
+  nextLink = '';
+  activityFeedContainer.innerHTML = '';
+};
+
+async function renderFeed() {
+  changeFilter();
+  populateActivityFeed({ category });
+  addIntersectionObserver();
+}
+
+const intersectionObserver = new IntersectionObserver(async (entries) => {
+  if (!nextLink) {
+    return;
+  }
+  if (entries[0].isIntersecting && !isDataLoading) {
+    await populateActivityFeed({ category }, nextLink);
+  }
+});
+
+const addIntersectionObserver = () => {
+  intersectionObserver.observe(lastElementContainer);
+};
+
+const removeIntersectionObserver = () => {
+  intersectionObserver.unobserve(lastElementContainer);
+};
+
+// main entry
+renderFeed();
+
+function renderActivityItem(data) {
+  const item = document.createElement('li');
+  item.classList.add('activity-item');
+  item.innerHTML = '';
+  let content = '';
+  switch (data.type) {
+    case logType.EXTENSION_REQUESTS:
+      content = formatExtensionRequestsLog(data);
+      break;
+    case logType.TASK:
+      content = formatTasksLog(data);
+      break;
+    case logType.REQUEST_CREATED:
+    case logType.REQUEST_APPROVED:
+    case logType.REQUEST_REJECTED:
+      content = formatOOORequests(data);
+      break;
+    default:
+      content = 'Unknown activity type';
+  }
+  item.innerHTML = `
+    ${content}`;
+  return item;
+}
+
+function formatOOORequests(data) {
+  let text = '';
+  if (data.type === logType.REQUEST_CREATED)
+    text = `<p><strong>${data?.user || data?.createdBy
+      }</strong> raised an OOO request from <strong>${new Date(
+        data.from,
+      ).toLocaleString()}</strong> until <strong>${new Date(
+        data.until,
+      ).toLocaleString()}</strong></p>`;
```

Comment:
- suggestion (optional)
  use tagged template literals to create respective components.